### PR TITLE
Changes TensorIterator computation to not consider out kwarg, lets UnaryOps safe cast to out

### PR DESCRIPTION
**Pull request by mruberry**
**BC breaking note:**

In PyTorch 1.5 passing the out= kwarg to some functions, like torch.add, could affect the computation. That is, 

```
out = torch.add(a, b)
```

could produce a different tensor than

```
torch.add(a, b, out=out)
```

This is because previously the out argument participated in the type promotion rules. For greater consistency with NumPy, Python, and C++, in PyTorch 1.6 the out argument no longer participates in type promotion, and has no effect on the computation performed. 

**ORIGINAL PR NOTE**

This PR effectively rewrites Tensor Iterator's "compute_types" function to both clarify its behavior and change how our type promotion works to never consider the out argument when determining the iterator's "common dtype," AKA its "computation type." That is, 

```
a = op(b, c)
```

should always produce the same result as

```
op(b, c, out=a)
```

This is consistent with NumPy and programming languages like Python and C++. 

The conceptual model for this change is that a TensorIterator may have a "common computation type" that all inputs are cast to and its computation performed in. This common computation type, if it exists, is determined by applying our type promotion rules to the inputs. 

A common computation type is natural for some classes of functions, like many binary elementwise functions (e.g. add, sub, mul, div...). (NumPy describes these as "universal functions.") Many functions, however, like indexing operations, don't have a natural common computation type. In the future we'll likely want to support setting the TensorIterator's common computation type explicitly to enable "floating ufuncs" like the sin function that promote integer types to the default scalar type. Logic like that is beyond the type promotion system, which can only review inputs.

Implementing this change in a readable and maintainable manner was challenging because compute_types() has had many small modifications from many authors over ~2 year period, and the existing logic was in some places outdated and in other places unnecessarily complicated. The existing "strategies" approach also painted with a broad brush, and two of them no longer made conceptual sense after this change. As a result, the new version of this function has a small set of flags to control its behavior. This has the positive effect of disentangling checks like all operands having the same device and their having the same dtype.

Additional changes in this PR:

- Unary operations now support out arguments with different dtypes. Like binary ops they check canCast(computation type, out dtype).
- The dtype checking for lerp was outdated and its error message included the wrong variable. It has been fixed.
- The check for whether all tensors are on the same device has been separated from other checks. TensorIterators used by copy disable this check.
- As a result of this change, the output dtype can be computed if only the input types are available.
- The "fast path" for checking if a common dtype computation is necessary has been updated and simplified to also handle zero-dim tensors.
- A couple helper functions for compute_types() have been inlined to improve readability. 
- The confusingly named and no longer used promote_gpu_output_dtypes_ has been removed. This variable was intended to support casting fp16 reductions on GPU, but it has become a nullop. That logic is now implemented here: https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/native/ReduceOpsUtils.h#L207.

### CODE REVIEW

File: aten/src/ATen/native/TensorIterator.h [hunk 1]
```diff
@@ -96,6 +96,10 @@ struct CAFFE2_API OperandInfo {
   // (e.g. if dtype is changed)
   Tensor original_tensor;
 
+  // If a tensor is both an input and an output, this lets the input
+  //   operand point to the output operand for a quick lookup
+  OperandInfo* corresponding_output = nullptr;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.h:101]: Is this an aliasing relation (if so, what if an input is multiple outputs) or a more technical definition (e.g., the self input in an inplace operation, which is both an input/output)?

File: aten/src/ATen/native/TensorIterator.h [hunk 2]
```diff
@@ -399,20 +397,23 @@ struct CAFFE2_API TensorIterator {
...
   bool check_mem_overlap_ = false;
   bool all_ops_same_shape_ = false;
   bool requires_channels_last_output_ = false;
   bool requires_channels_last_3d_output_ = false;
   bool static_shape_ = false;
+  bool check_all_same_dtype_ = true;
+  bool check_all_same_device_ = true;
+  bool promote_inputs_to_common_dtype_ = false;
+  bool cast_common_dtype_to_outputs_ = false;
+  bool enforce_safe_casting_to_output_ = false;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.h:416]: What is "safe casting to output"?

File: aten/src/ATen/native/cpu/LerpKernel.cpp [hunk 3]
```diff
@@ -38,11 +35,8 @@ static void lerp_kernel_tensor(
     const Tensor& self,
     const Tensor& end,
     const Tensor& weights) {
-  // lerp() only uses TensorIterator for CPU. Since TensorIterator would
-  // would attempt to promote types inconsistent with the CUDA implementation,
-  // restrict types explicitly here
   TORCH_CHECK(self.dtype() == end.dtype(), "expected dtype ", self.dtype(), " for `end` but got dtype ", end.dtype());
-  TORCH_CHECK(self.dtype() == weights.dtype(), "expected dtype ", self.dtype(), " for `weights` but got dtype ", end.dtype());
```
Comment (ezyang) [inline aten/src/ATen/native/cpu/LerpKernel.cpp:45]: lol

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 4]
```diff
@@ -212,7 +212,7 @@ static TensorIterator make_reduction(
 
 static TensorIterator make_reduction(
     const char* name, Tensor& result1, Tensor& result2, const Tensor& self, IntArrayRef dim,
-    bool keepdim, ScalarType dtype1, ScalarType dtype2, bool promote_gpu_output_dtypes=true)
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOpsUtils.h:215]: Do you know the historical context for this argument?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 5]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
   permute_dimensions(perm_);
 }
 
-// Returns the first non-CPU device, if present, or the CPU
-// if all operands are on the CPU.
-Device compute_device(at::ArrayRef<OperandInfo> operands) {
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:78]: Appears to have been folded into `compute_types`

File: aten/src/ATen/native/TensorIterator.cpp [hunk 6]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
-      }
-    } else {
-      all_same_type = false;
-      break;
-    }
-  }
-  if (all_same_type) {
-    return std::make_tuple(device, common_type, true);
+// See [Result type computation] in TensorIterator.h
+// NOTE: This is (or at least, was) slow, and compute_types() (below) attempts
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:77]: "was" --> so what was this previously? `compute_common_type_`?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 7]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+ScalarType TensorIterator::compute_common_dtype() {
+  if (operands_.size() - num_outputs_ == 0) {
+    return ScalarType::Undefined;
   }
 
   at::native::ResultTypeState state = {};
-  for (const auto& op : operands) {
+  for (const auto& op : operands_) {
+    if (op.is_output) {
+      continue;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:87]: This is the main substantive change of this PR right?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 8]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
-    }
-  }
-  if (all_same_type) {
-    return std::make_tuple(device, common_type, true);
+// See [Result type computation] in TensorIterator.h
+// NOTE: This is (or at least, was) slow, and compute_types() (below) attempts
+//   to avoid using it.
+ScalarType TensorIterator::compute_common_dtype() {
+  if (operands_.size() - num_outputs_ == 0) {
+    return ScalarType::Undefined;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:81]: Shouldn't you also set `common_dtype_` here?

Comment (ezyang) [reply aten/src/ATen/native/TensorIterator.cpp:81]: (What case is this optimization for?)

File: aten/src/ATen/native/TensorIterator.cpp [hunk 9]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+//     - check_all_same_dtype_, if true all operands must have the same dtype
+//     - check_all_same_device_, if true all operands must be on the same device
+//     - promote_inputs_to_common_dtype_, if true allows temporaries to be
+//         created for inputs on the CPU and sets input operands' target
+//         dtypes to the computation dtype
+//     - cast_common_dtype_to_outputs_, if true allows temporaries to be
+//         created to hold the results of the computation on the CPU and
+//         sets output operands' target dtypes to the computation dtype
+//     - enforce_safe_casting_to_output_, if true
+//         canCast(computation_dtype, output dtype) must be true
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:119]: OK. Turn this comment into a Note and then refer to it from the members

File: aten/src/ATen/native/TensorIterator.cpp [hunk 10]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+//     (3) check that all operands share the same dtype, if requested
+//     (4) create temporaries as inputs and outputs, if needed and requested,
+//           on the CPU
+//   The CPU-specific behavior is because CUDA kernels are expected to
+//   cast to and from the computation dtype in the kernel itself.
+//
+//   Several flags control this behavior, they are:
+//     - check_all_same_dtype_, if true all operands must have the same dtype
+//     - check_all_same_device_, if true all operands must be on the same device
+//     - promote_inputs_to_common_dtype_, if true allows temporaries to be
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:112]: Shouldn't this explicitly say "cpu" in its name then?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 11]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+//         created for inputs on the CPU and sets input operands' target
+//         dtypes to the computation dtype
+//     - cast_common_dtype_to_outputs_, if true allows temporaries to be
+//         created to hold the results of the computation on the CPU and
+//         sets output operands' target dtypes to the computation dtype
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:117]: I don't know what this means. What's an operand "target dtype"? Isn't it more of a "source" dtype?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 12]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+//     - enforce_safe_casting_to_output_, if true
+//         canCast(computation_dtype, output dtype) must be true
+//
+//   NOTE: Checks for more specific behaviors (e.g. the first and second
+//     inputs must share a dtype, but the third must have the long dtype)
+//     should be implemented by users.
+//   NOTE: users should not requet the iter's common dtype without
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:124]: request

File: aten/src/ATen/native/TensorIterator.cpp [hunk 13]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+//     checking that all dtypes are the same or enabling type promotion.
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:125]: Can you set a `TORCH_INTERNAL_ASSERT` to test for this invariant?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 14]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+void TensorIterator::compute_types() {
+  // Reviews operands (1/2)
+  //   - sets original tensor
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:128]: It's weird that this isn't done in the operand constructor

File: aten/src/ATen/native/TensorIterator.cpp [hunk 15]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+                        (has_undefined_outputs || enforce_safe_casting_to_output_ ||
+                        cast_common_dtype_to_outputs_)));
+
+  // Computes a common dtype, if needed
+  // NOTE: only occurs if there are multiple distinct dtypes
+  if (has_different_input_dtypes && promote_inputs_to_common_dtype_) {
+    common_dtype_ = compute_common_dtype();
+  }
+
+  // Reviews operands (2/2) - CUDA and CPU cases
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:205]: I'm still trying to understand the new code, but I'm not a big fan of having copy pasted the code for the cpu / cuda cases (with the behavior changes you described, ofc)

File: aten/src/ATen/native/TensorIterator.h (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.h:101]: I also don't like this raw pointer. I'll get rid of it and fold the aliasing check into the logic which reuses the temporary/ies.

File: aten/src/ATen/native/TensorIterator.h (hunk 2 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.h:416]: Your plan to make the comment block in TensorIterator.cpp a note and reference it from here should clarify. I agree it's not clear.

File: aten/src/ATen/native/ReduceOpsUtils.h (hunk 4 above)
Comment (mruberry) [reply aten/src/ATen/native/ReduceOpsUtils.h:215]: Yes. I added a brief reference for it in the PR note. I reviewed with @ngimel and the team at NVIDIA before removing it.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 6 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:77]: I should improve the clarity of this comment. @ngimel suggested avoiding this path because it's been slow in the past, but the speed may have improved.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 7 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:87]: Conceptually? Yes.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 8 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:81]: We probably don't need this short-circuit anymore, but for clarity we could set the common_dtype there, sure. It's already set to ScalarType::Undefined.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 9 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:119]: Good idea!

File: aten/src/ATen/native/TensorIterator.cpp (hunk 10 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:112]: I went back and forth on this myself, but although the behavior on CUDA is more subtle it applies to both. It disables the check that all tensors are the same dtype on both (1) and sets the target dtype on both (2).

File: aten/src/ATen/native/TensorIterator.cpp (hunk 11 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:117]: That's a great question. I can probably add some additional commentary clarifying this.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 13 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:125]: Good idea. It will require fixing a few more functions (analogous to lerp), but they should be fixed, anyway.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 14 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:128]: Agreed. I think I can move it there. I'll check.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 15 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:205]: Me either. It may be a reactionary response to the previous code, which was very hard to parse with a plethora of cases. The actual lines of overlap is smaller than it appears. I'll take another look at merging them and if it doesn't impact the readability much I'll do it.

**Review by ezyang — APPROVED** [review]
Approving in the spirit of moving things along.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 16]
```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
...
+    }
+
+    // Casts to outputs by creating temporaries of the correct dtype (if needed)
+    // NOTE: for inplace operations this reuses the copy made when promoting
+    //   inputs, when available
+    if (op.is_output && op.current_dtype != common_dtype_ && cast_common_dtype_to_outputs_ &&
+        !(op.is_read_write && promote_inputs_to_common_dtype_)) {
+      op.tensor = at::empty_like(op.tensor,
+                                 op.tensor.options().dtype(common_dtype_),
+                                 LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
Comment (ngimel) [inline aten/src/ATen/native/TensorIterator.cpp:323]: Why not preserve current output format?

File: test/test_type_promotion.py [hunk 17]
```diff
@@ -69,7 +69,7 @@ def test_inplace(self, device):
         uint8_tensor *= int16_tensor
 
     @float_double_default_dtype
-    def test_unsinged(self, device):
```
Comment (ngimel) [inline test/test_type_promotion.py:72]: lol

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/TensorIterator.h (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.h:101]: I removed this optimization to avoid complicating the code and dragging down the common path with additional operand enumerations.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 16 above)
Comment (mruberry) [reply aten/src/ATen/native/TensorIterator.cpp:323]: Trying to preserve as much of the previous behavior as possible.
